apiserver/facades/client/cloud/cloud_test.go

```diff
@@ -824,15 +1031,38 @@ func (s *cloudSuite) TestUpdateCredentialsModelFailedValidation(c *gc.C) {
 }
 
 func (s *cloudSuite) TestUpdateCredentialsModelFailedValidationForce(c *gc.C) {
-	s.backend.credentialModelsF = func(tag names.CloudCredentialTag) (map[string]string, error) {
-		return map[string]string{
-			coretesting.ModelTag.Id(): "testModel1",
-		}, nil
+	adminTag := names.NewUserTag("admin")
+	defer s.setup(c, adminTag).Finish()
+
+	aCloud := jujucloud.Cloud{
+		Name:      "dummy",
+		Type:      "dummy",
+		AuthTypes: []jujucloud.AuthType{jujucloud.EmptyAuthType, jujucloud.UserPassAuthType},
+		Regions:   []jujucloud.Region{{Name: "nether", Endpoint: "endpoint"}},
 	}
 
-	s.PatchValue(cloudfacade.ValidateNewCredentialForModelFunc, func(backend credentialcommon.PersistentBackend, callCtx context.ProviderCallContext, credentialTag names.CloudCredentialTag, credential *cloud.Credential, migrating bool) (params.ErrorResults, error) {
-		return params.ErrorResults{[]params.ErrorResult{{&params.Error{Message: "not valid for model"}}}}, nil
-	})
+	model := map[string]string{
+		coretesting.ModelTag.Id(): "testModel1",
+	}
+
+	s.PatchValue(cloudfacade.ValidateNewCredentialForModelFunc,
+		func(
+			backend credentialcommon.PersistentBackend, callCtx context.ProviderCallContext,
+			credentialTag names.CloudCredentialTag, credential *jujucloud.Credential, migrating bool,
+		) (params.ErrorResults, error) {
+			return params.ErrorResults{[]params.ErrorResult{{&params.Error{Message: "not valid for model"}}}}, nil
+		})
```

Comment:
There are similar `PatchValue` in a lot of places. This may not be possible given the slightly different behaviours, but would it be possible to set this up in SetupTest?

Reply:
```
s.PatchValue(cloudfacade.ValidateNewCredentialForModelFunc,
		func(
			_ credentialcommon.PersistentBackend, _ context.ProviderCallContext,
			_ names.CloudCredentialTag, _ *jujucloud.Credential, _ bool,
		) (params.ErrorResults, error) {
			return params.ErrorResults{[]params.ErrorResult{{&params.Error{Message: "not valid for model"}}}}, nil
		})
```
Actually observed that this type of `Patches` that just return an error are not required as long as the Assert is asserting for the error. So I have removed all occurences of this `PatchValue` and only left with the ones that patch the mock method to actually test for an actual behaviour.

Reply:
Also I think I would rather let the patch be there as it is easy to see what function it is patching and its immediate effect to the test that it is used. As opposed to setting some patch method calls in setup, and have the reader making that back and forth look up into setup to remember what function is being patched.